### PR TITLE
MASSEMBLY-919: Fix dependency scope handling

### PR DESCRIPTION
**Pull request by gzsombor**
The problem is that the assembly plugin collects every dependency, flatten the dependency hierarchy tree, and only after this, filter's out by scope. This will cause that if A has a test dependency B, which has a 'runtime' dependency C, C is treated as a runtime dependency of A, and it is included. The implemented fix is using proper tree walking, and checking eagerly the scope of the dependency.

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [X] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MASSEMBLY) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [X] Format the pull request title like `[MASSEMBLY-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MASSEMBLY-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [X] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [ ] You have run the integration tests successfully (`mvn -Prun-its clean verify`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [X] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

File: src/main/java/org/apache/maven/plugins/assembly/artifact/ScopeBasedDependencySelector.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+    private boolean providedScope;
+
+    private boolean systemScope;
+
+    public ScopeBasedDependencySelector( String scope )
+    {
+        switch ( scope )
+        {
+            case Artifact.SCOPE_PROVIDED:
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/assembly/artifact/ScopeBasedDependencySelector.java]: This looks like there's a missing type somewhere, perhaps an enum, which would have these values as properties

File: src/main/java/org/apache/maven/plugins/assembly/artifact/ScopeBasedDependencySelector.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+                runtimeScope = true;
+                testScope = true;
+                break;
+            default:
+                throw new RuntimeException( "Unknown scope specified:" + scope );
+        }
+    }
+
+    @Override
+    public boolean selectDependency ( Dependency dependency )
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/assembly/artifact/ScopeBasedDependencySelector.java]: name feels off given that this returns a boolean

**Comment by struberg** [comment]
@gzsombor can you please help us draft an integration test first?
I tried to replay your described scenario and I do not see any errors.
Here is the code https://github.com/struberg/maven-assembly-plugin/tree/MASSEMBLY-919

Where does my IT differ from your use case?

**Comment by rfscholte** [comment]
I've replaced Easymock with Mockito, so this PR needs to be updated

**Comment by Geker** [comment]
any update ?
this problem is still exists

**Comment by cstamas** [comment]
Ideally we need an IT (or some test) that reproduces the issue (fails the expectation) and then a PR that fixes it (makes IT pass).

@Geker can you provide a reproducer? or does https://github.com/apache/maven-assembly-plugin/pull/12#pullrequestreview-336128503 work?

**Comment by slawekjaranowski** [comment]
We should direct use Resolver API.
By the way thanks for effort.

**Comment by jira-importer** [comment]
Resolve #1129

**Pull request closed without merging**